pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -605,25 +608,109 @@ private static File convertURIToSegmentLocation(String segmentLocation) {
    */
   public boolean commitSegmentMetadata(String rawTableName, final String committingSegmentNameStr, long nextOffset,
       long memoryUsedBytes) {
-    final long now = System.currentTimeMillis();
     final String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
 
+    /*
+     * Update zookeeper in 3 steps.
+     *
+     * Step 1: Update PROPERTYSTORE to change the old segment metadata status to DONE
+     * Step 2: Update PROPERTYSTORE to create the new segment metadata with status IN_PROGRESS
+     * Step 3: Update IDEALSTATES to include new segment in CONSUMING state, and change old segment to ONLINE state.
+     *
+     * The controller may fail between these three steps.
+     * So when a new controller takes over as leader, or when validation manager runs, it needs to check the following:
+     *
+     * If it fails between step-1 and step-2:
+     * Check whether there are any segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS,
+     * and hence the status of the segment in the IDEALSTATE is still CONSUMING
+     *
+     * If it fails between step-2 and-3:
+     * Check whether there are any segments in PROPERTYSTORE with status IN_PROGRESS, that are not accounted for in idealState.
+     * If so, it should create the new segments in idealState.
+     *
+     * If the controller fails after step-3, we are fine because the idealState has the new segments.
+     * If the controller fails before step-1, the server will see this as an upload failure, and will re-try.
+     */
+
+    final LLCSegmentName committingLLCSegmentName = new LLCSegmentName(committingSegmentNameStr);
+
+    // Step-1
+    boolean success = updateOldSegmentMetadataZNRecord(realtimeTableName, committingLLCSegmentName, nextOffset);
+    if (!success) {
+      return false;
```

Comment:
what happens on the server side, if this method returns false? Will the servers retry?

Reply:
the servers handle receiving false just fine. they retry to the appropriate lead controller again.

Reply:
I thought you mentioned that all retries will fail because metadata is already updated. In this case IdealState will not be updated until validation manager kicks in. Until then the server will keep receiving false and keep retrying? 
. @npawar I am trying to understand the flow. Feel free to merge